.gitlab-ci.yml

```diff
@@ -1,40 +1,91 @@
-# default build for all distros
-# only create artifacts of the build directory when something fails (for cmake logs)
-# cache the ccache/ directory for each job separately
-.build_template: &distro_build
+# all builds use the same ccache folder in the project root that is cached
+variables:
+  CCACHE_BASEDIR: '$CI_PROJECT_DIR'
+  CCACHE_DIR: '$CI_PROJECT_DIR/ccache'
+
+# default config for all distros:
+# - install dependencies via script
+# - create ccache dir & setup caching of it (for each job separately)
+.build_config: &default_config
   before_script:
     - ci/install_dependencies.sh
+    - mkdir -p ccache
+  cache:
+    key: "$CI_JOB_NAME"
+    paths:
+      - ccache/
+
+# default build job:
+# - run build script
+# - only create artifacts of the build directory when something fails
+#   (for cmake logs)
+.build_template: &distro_build
   script:
     - python3 ci/test_build.py
   artifacts:
     when: on_failure
     paths:
       - build/
-  cache:
-    key: "$CI_JOB_NAME"
-    paths:
-      - ccache/
+
+stages:
+  - test
+  - deploy
 
 Fedora:
   image: fedora:28
+  <<: *default_config
   <<: *distro_build
 
 Debian:
   image: debian:9
+  <<: *default_config
   <<: *distro_build
 
 Archlinux:
   image: base/archlinux
+  <<: *default_config
   <<: *distro_build
 
 Ubuntu:
   image: ubuntu:18.04
+  <<: *default_config
   <<: *distro_build
 
 CentOS:
   image: centos:7
+  <<: *default_config
   <<: *distro_build
 
 OpenSUSE:
   image: opensuse:tumbleweed
+  <<: *default_config
   <<: *distro_build
+
+Install:
+  stage: deploy
+  image: fedora:28
+  <<: *default_config
+  script:
+    - mkdir build && cd build
+    - cmake -DCMAKE_BUILD_TYPE=Release -DEXIV2_ENABLE_VIDEO=ON -DEXIV2_ENABLE_WEBREADY=ON -DEXIV2_BUILD_UNIT_TESTS=ON -DCMAKE_INSTALL_PREFIX=/usr/ -DBUILD_WITH_CCACHE=ON ..
```

Comment:
[suggestion] Could we install the project in a local folder inside the build directory (`build/install`) ? I do not like the idea of installing things directly in `/usr/` even when we are throwing away these CI images after every run. I am also surprised this is working without requiring any `sudo` command ... Does this means that all the commands are executed as **root**?

Reply:
Yes, the whole thing is running as root, otherwise you wouldn't be able to install packages. But I though that was the general approach for docker containers anyway.

Concerning the location, I was trying to simulate an actual installation, that puts all the binaries into `$PATH`. It won't probably make a difference though where we install anything.

Reply:
I tried changing the directory directory to something inside the repository, but that leads to all kinds of wonky problems. The biggest one is that `libexiv2.so` is then no longer in the ld library path and to get the dynamic linker to find it we have to either modify `LD_PRELOAD` or mess with it's configs. To be honest I'd rather let it install to `/usr/` as it works and that's the default target anyway (I guess).

Reply:
Oh, I see. Let's get this merged as it is and I can investigate later the issue with the RPATHs. With CMake we can do some magic to edit the RPATHs at the installation point to fix this kind of issues. I'll create an issue to not forget about this.